[docs] Add documentation for the font plugin (#25754)
expo/expo: docs/pages/develop/user-interface/fonts.mdx

## Use a custom font

> There are two ways to use a custom font in your project: you can either embed the font in your native project or load it at runtime. The first approach is recommended as it is simpler and more reliable. The second approach is useful if you want to load the font in Expo Go / without creating a new native build that includes it, or load it from a remote URL.

Reviewer: ```suggestion
There are two ways to use a custom font in your project: either embed the font in your native project or load it at runtime. The first approach is recommended as it is simpler and more reliable. The second approach is useful if you want to load the font in Expo Go/without creating a new native build that includes it or loading it from a remote URL.
```

> ### 1. Embed the font in your native project

Reviewer: ```suggestion
### Embed the font in your native project
```

This seems like either/or approach so let's not use numbers in headers as it causes confusion if the steps are continuous. If my understanding is wrong here, let me know and I will update these sections to use Step component.

> > Note: this `expo-font` config plugin is only available for SDK 50 and higher. If you are using an older SDK, you can [load the font at runtime](#2-load-the-font-at-runtime) instead. 

Reviewer: ```suggestion
> **Note:** The `expo-font` config plugin is only available for SDK 50 and above. If you are using an older SDK, you can [load the font at runtime](#load-the-font-at-runtime) instead. 
```

> Install [`expo-font`](/versions/latest/sdk/font/#installation) package and add the config plugin to your `app.json` file. The plugin will embed the font in your native project.

Reviewer: ```suggestion
Install [`expo-font`](/versions/latest/sdk/font/#installation) library and add the config plugin to your [app config](/versions/latest/config/app/) file. The plugin will embed the font in your native project.
```

- Let's prefer library instead of package as per our writing style guide
- Let's link to app.json config and use app config because app.json file can also be app.config.js, etc.

> ```json

Reviewer: ````suggestion
```json app.json
````

{
  "expo": {
    "plugins": [
      [
        "expo-font",
        {
          "fonts": [
            "./assets/fonts/Inter-Black.otf"
          ]
        }
      ]
    ]
  }
}
```

> 

Reviewer: ```suggestion

The `fonts` option takes an array of one or more font files to link to the native project. The path to each font file is relative to the project's root.

```

It wasn't obvious to me until went through the changes in API reference that this array can take multiple docs. Maybe it is just me so feel free to ignore this but stating the obvious from the above one liner might help avoid this confusion.

> [Create a new native build](/develop/development-builds/create-a-build/), and you can now use the font in your project with the `fontFamily` style prop. The font family name will be the name of the font file, without the extension. For example, if you have a font file named **Inter-Black.otf**, the font family name will be **Inter-Black**.

Reviewer: ```suggestion
[Create a new native build](/develop/development-builds/create-a-build/), and you can now use the font in your project with the `fontFamily` style prop. The font family name will be the name of the font file without the extension. For example, if you have a font file named **Inter-Black.otf**, the font family name will be **Inter-Black**.
```

> If you'd like to try out a font without creating a new build, you could alternatively load it at runtime. This is what the second approach explains.

Reviewer: ```suggestion
If you want to try a font without creating a new build, you can load it at runtime. This is what the next section explains.
```

<ConfigReactNative>

- **Android:** Copy font files to **android/app/src/main/assets/fonts**.

> - **iOS**: See ["Adding a Custom Font to Your App" on the Apple Developer documentation](https://developer.apple.com/documentation/uikit/text_display_and_fonts/adding_a_custom_font_to_your_app).

Reviewer: ```suggestion
- **iOS**: See [Adding a Custom Font to Your App](https://developer.apple.com/documentation/uikit/text_display_and_fonts/adding_a_custom_font_to_your_app) in the Apple Developer documentation.
```

</ConfigReactNative>

> ### 2. Load the font at runtime

Reviewer: ```suggestion
### Load the font at runtime
```

> After getting the font file, in your project, you need to install [`expo-font`](/versions/latest/sdk/font/#installation) package.

Reviewer: ```suggestion
After getting the font file, in your project, you need to install [`expo-font`](/versions/latest/sdk/font/#installation) library.
```

Outside the scope of this PR but let's also use library instead of package here. If you think introducing this change here is not worth it, I am happy to audit this doc in a follow up PR.

<Terminal cmd={['$ npx expo install expo-font @expo-google-fonts/inter']} />

> Then, you can integrate it in your project with the config plugin by passing in the path to the desired font file to the `fonts` option (eg: `node_modules/@expo-google-fonts/inter/Inter_100Thin.ttf`), or you can use the `useFonts` hook.

Reviewer: ````suggestion
Then, you can [embed it into your project with the config plugin](#embed-the-font-in-your-native-project) by passing the path to the desired font file to the `expo-font.fonts` option, or you can use the `useFonts` hook.

An example of the path provided to the `fonts` option in the config plugin:

```json app.json
{
  "plugins": [
    [
      "expo-font",
      {
        "fonts": "node_modules/@expo-google-fonts/inter/Inter_100Thin.ttf"
      }
    ]
  ]
}
```
````

- Let's add link to the config plugin section.
- Let's also use the same terminology (embed) instead
- Also, provide the example of provide for the `node_modules` path too

> The Google Fonts package provides the `useFonts` hook for convenience. Under the hood, the hook uses [`Font.loadAsync`](/versions/latest/sdk/font/#loadasyncfontfamilyorfontmap-source). You do not have to explicitly import the font file since that is done by the package itself.

Reviewer: ```suggestion
The Google Fonts package provides the `useFonts` hook for convenience. Under the hood, the hook uses [`Font.loadAsync`](/versions/latest/sdk/font/#loadasyncfontfamilyorfontmap-source). You do not have to explicitly import the font file since the package itself does that.
```

## Wait for fonts to load

> If you [embed your fonts in your native project](#1-embed-the-font-in-your-native-project), they will be available immediately without any additional code outside of the config plugin. However, if you load your fonts at runtime, they will not be available immediately — so, it is generally a good practice to not render anything until the font is ready. Instead, you can continue to display the Splash Screen of your app until all fonts have loaded (or an error has been returned). It is done by using [`expo-splash-screen`](/versions/latest/sdk/splash-screen/) package. See the [minimal example](#minimal-example) section on how to use it.

Reviewer: ```suggestion
If you [embed your fonts in your native project](#embed-the-font-in-your-native-project), they will be available immediately without any additional code outside the config plugin. However, if you load your fonts at runtime, they will not be available immediately. So, it is generally a good practice not to render anything until the font is ready. Instead, you can continue to display the Splash Screen of your app until all fonts have loaded (or an error has been returned). It is done using [`expo-splash-screen`](/versions/latest/sdk/splash-screen/) package. See the [minimal example](#minimal-example) section on how to use it.
```

Let's use a proper line break instead here.

### Load fonts on the web

Sometimes, particularly on the web people choose to render their content in a platform default font while their custom font is loading. Alternatively, to render the rest of their content, that doesn't depend on the custom font while the font is loading. These approaches are called FOUT and FOIT and you can read a lot more about them on the web.

In general, these strategies are not recommended for native apps. If you include your fonts in your project, the fonts will always be delivered to the user by the time your code is running. The one exception to this is that you may prefer to do this on the web.

## Additional information

You probably don't need to know anything beyond this point to use custom fonts effectively in your app. If you are curious or your use case has not been addressed by the above information, please continue reading.

> ### Loading a remote font directly from the web

Reviewer: In the whole document, we're using short-form verbs in the headers for consistency instead of gerunds, so I'm not sure why it is introduced for this and the last sub-section.

Author: the problem with saying "Load a remote font directly from the web" is that it sounds like we're recommending that people do this, rather than explaining how you can do this

Reviewer: I understand it now. Thanks for providing the context 🙏
